StatesDumper/StatesDumper.cs

```diff
@@ -63,5 +65,77 @@ private bool OnHelp(string[] args)
             Console.Write($"{Name} Commands:\n" + "\tdump storage <key>\n");
             return true;
         }
+
+        public void OnPersist(Snapshot snapshot)
+        {
+            switch (Settings.Default.PersistAction)
+            {
+                // Action == 0 calls OnPersistStorage
+                case 0:
+                    OnPersistStorage(snapshot);
+                    break;
+            }
+        }
+
+        private static void OnPersistStorage(Snapshot snapshot)
+        {
+            uint blockIndex = snapshot.Height;
+            if (blockIndex >= Settings.Default.HeightToBegin)
+            {
+                string dirPath = "./Storage";
+                Directory.CreateDirectory(dirPath);
+                string path = $"{HandlePaths(dirPath, blockIndex)}/dump-block-{blockIndex.ToString()}.json";
+
+                JArray array = new JArray();
+
+                foreach (DataCache<StorageKey, StorageItem>.Trackable trackable in snapshot.Storages.GetChangeSet())
+                {
+                    JObject state = new JObject();
+
+                    switch (trackable.State)
+                    {
+
+                        case TrackState.Added:
+                            state["state"] = "Added";
+                            state["key"] = trackable.Key.ToArray().ToHexString();
+                            state["value"] = trackable.Item.ToArray().ToHexString();
+                            // Here we have a new trackable.Key and trackable.Item
+                            break;
+                        case TrackState.Changed:
+                            state["state"] = "Changed";
+                            state["key"] = trackable.Key.ToArray().ToHexString();
+                            state["value"] = trackable.Item.ToArray().ToHexString();
+                            break;
+                        case TrackState.Deleted:
+                            state["state"] = "Deleted";
+                            state["key"] = trackable.Key.ToArray().ToHexString();
+                            break;
+                    }
+                    array.Add(state);
+                }
+
+                Settings.Default.BlockStorageCache = Settings.Default.BlockStorageCache + "{\"block\":" + blockIndex.ToString() + ",\"size\":" + array.Count.ToString() + ",\"storage\":\n";
+                Settings.Default.BlockStorageCache = Settings.Default.BlockStorageCache + array.ToString() + "},\n";
+
+                if ((blockIndex % Settings.Default.BlockCacheSize == 0) || (blockIndex > Settings.Default.HeightToStartRealTimeSyncing))
```

Comment:
Does this prevent data loss?

Reply:
I think the only way to prevent data loss is to save the data every time the block is persisted.

Reply:
It does not prevent, @erikzhang , you are right.

What we did is the option of choosing the `BlockCacheSize`.
If 'BlockCacheSize==1` there will be no data loss.

In our case, we chose it `BlockCacheSize==1000` (it is a little faster instead of persisting every block when syncing for the first time) and dumped with `ImportPlugin chain.acc`.

The chance of data loss is lower when Importing.
That is why there is an additional variable that is `HeightToStartRealTimeSyncing` which, implicitly, forces BlockCacheSize to 1.

There could be a tool that prints it every block and later deletes the files and merge then.
We currently do it in `bash` scripts: https://github.com/NeoResearch/neo-tests/blob/master/simple-rpc-node/merge_Storage.sh

Reply:
What's the difference between `BlockCacheSize=1` and `BlockCacheSize=1000`? Is it related to performance?

Reply:
The performance was not good when persisting file on every block during import, then we created this possible cache.
What do you think, @igormcoelho?

Reply:
Yeah, reducing disk writes is fundamental for performance... we experimented with BlockCacheSize=1 and it was very slow during block import... the bigger the Cache the better. We found that 1000 was a good compromise between risking losing block loading and offloading it to disk at every few seconds (during block import only, after that behavior should go to 1 again, which is enough for every 15 seconds).